### DISCUSSION TITLE
How to import mode source from other modesolver

### DISCUSSION THREAD

**Discussion by 694856790** [NanoComp/meep#2574] [Q&A]
Hello, I want to know whether MEEP can support external mode field import, because I found that MPB can not calculate the mode field of lossy media, so I want to obtain the mode field through other mode solver to use as the mode source, is there a good way? Thanks.

**Comment by smartalecH** [comment]
Sure if you have the mode field profile (eg computed with a different mode solver), you can inject that as a source. Here's a thread you can follow:

https://github.com/NanoComp/meep/discussions/2218#discussioncomment-3512978

@hammy4815 did you write up a tutorial that does this? @oskooi  might be good to add this as a FAQ.

**Reply by 694856790** [reply]
Thank you for your reply. It is precisely because I saw the issue of # 2218 that I raised this question. May I ask if the mode source obtained through the equivalent source method is suitable for 3D meep calculations? And most importantly, is the mode source emitting in both directions or in one direction like the Eigenmode source?

**Reply by smartalecH** [reply]
If you provide the E and H components, it will only propagate in one direction. Yes it should work in 3D.

**Reply by 694856790** [reply]
Thank you for your quick reply! I would like to know if there is a way to use a mode monitor if this is done, as my ultimate goal is to obtain the S parameter of an optical device composed of a lossy medium.

**Reply by ianmatthewhammond** [reply]
I have not made a tutorial, but I can add this to my ToDo. As for the 3D @694856790 , yes it works in 3D provided the source is a plane (2D image). You can still use mode monitors in MEEP exactly as you would along with any other source using `add_mode_monitor` and `get_eigenmode` as described in the docs.

**Reply by 694856790** [reply]
@hammy4815 Thanks! The mode source is a 2D image ( a 2D matrix), which I think should match your description. However, I am not very familiar with the pattern monitor. Can I load my own mode fields into the mode monitor to get the mode coefficient to calculate the S parameters?

**Reply by ianmatthewhammond** [reply]
Ah, if you want to calculate the mode coefficients using your provided mode patterns, you can just use `get_dft_array` to grab the field profile at a given frequency and manually compute the mode decomposition using [overlap integrals (see here)](https://meep.readthedocs.io/en/latest/Mode_Decomposition/#how-mode-decomposition-works).

**Reply by 694856790** [reply]
That's a good idea, thank you!

**Reply by 694856790** [reply]
@hammy4815 By the way, I want to confirm, should I use a normal DFT monitor or a mode monitor to capture the profile of fields? The mode monitor should distinguish between energy forward transmission and backward transmission, is that right?

**Reply by ianmatthewhammond** [reply]
You should just use the DFT monitor if you want it to match your provided mode profiles, rather than those solved for by MPB. To distinguish between forward and backward propagating modes, you need both profiles to compute their respective overlaps.
